source/common/router/router.cc

```diff
@@ -876,6 +893,10 @@ void Filter::UpstreamRequest::onPoolReady(Http::StreamEncoder& request_encoder,
     parent_.downstream_headers_->Host()->value(host->hostname());
   }
 
+  if (span_ != nullptr) {
+    span_->injectContext(*parent_.downstream_headers_);
```

Comment:
If I have understood correctly, this new child span would be the parent of the client spans representing the actual call to the destination service? So if there are multiple retries, then there would be a client span per retry, but they would all be contained within this new child span?

If this is the case - then I don't think you need to inject the context - that is only relevant for client spans to propagate the context to the invoked service.

Reply:
I might be mistaken, but I think we do need the inject context here. The idea is that Envoy already creates a parent span on HTTP connection manager ingress. In the current code, we always just forward that context through the router. However, if we make multiple calls out in the context of the original ingress span (e.g., auth, ratelimit, etc.) it's unclear how the calls line up. My goal here is to have an option where every routed call creates a new span. This way we see something like:

```
-> ingress
  -> ratelimit
  -> auth
  -> ratelimit
  -> router
    -> some service ingress
```

I assume in this case I need to inject context on each outbound call? In the case of retry with this code we would see something like:

```
-> ingress
  -> ratelimit
  -> auth
  -> ratelimit
  -> router
    -> some service ingress
  -> router
    -> some service ingress
```

Right now there are no tags that indicate retry, etc. but I will add those also in a follow up.

Reply:
Ok think I understand now.

Currently the default tracing shows spans for the inbound and outbound requests on behalf of the proxied service - but the additional tracing in your example shows internal spans representing the additional work performed by the proxy, including routing to the local service.

Was wondering whether, instead of specific config properties (e.g. start_child_span) whether a higher level config property could just enable all the internal tracing for the proxy?

Reply:
> Was wondering whether, instead of specific config properties (e.g. start_child_span) whether a higher level config property could just enable all the internal tracing for the proxy?

I thought about this, but I think it has to be per-router because some proxies are doing both ingress and egress, and I think having a double-span in some flows would be noisy. I think it's possible that we could trigger this off of the HTTP connection manager config based on ingress/egress, but even that is messy because for an edge/middle proxy we are doing both ingress/egress in the same listener. Thoughts?

Reply:
Not sure how the tracing would work if same listener used for both anyway.

But i think it would be better if could be managed by combination of tracing config ingress/egress property and another property indicating if proxy internals should also be traced.

Reply:
> Not sure how the tracing would work if same listener used for both anyway.

It definitely works. This is standard on edge (middle) proxy nodes. The same listener is both ingress/egress. Let me think a bit more as to whether it makes sense to derive this from the parent HTTP connection manager ingress/egress setting.

Reply:
@objectiser let's sync up when you are back from vacation and we also have the email thread going. I've come full circle on this and given the range of situations we need to support I now think that the granular config option on the router is the way to go. (So basically I would like to keep this approach).

Reply:
Alright per offline discussion I think we are OK to merge this as is. @objectiser WDYT?